File: 프로그래머스 풀이/lv2_주식가격.java
import java.util.*;
class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];    
        
        Stack<Integer> pri = new Stack<>();
        Stack<Integer> index = new Stack<>();
        int time = 0;
        pri.push(prices[0]);
        index.push(0);
        
        int idx = 0;
        for (int i = 1; i < prices.length; i++) {
            time++;
            while(pri.empty() == false && pri.peek() > prices[i]) {        
                idx = index.pop();
                answer[idx] = time-idx;
                pri.pop();          
            }
            pri.push(prices[i]);
            index.push(i);
        }
        while (pri.empty() == false) {
            idx = index.pop();
            answer[idx] = time-idx;
            pri.pop();
        }
        return answer;
    }
}

File: 프로그래머스 풀이/lv2_주식가격.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> prices) {
    int len = prices.size();
    vector<int> answer(len, -1);
    int i, j, top=0, t, time=0, pri;
    int stack[len];    // idx 넣기(시간)
    stack[0] = 0;
    
    for (i = 1; i < len; i++) {
        pri = prices[i];
        if ( top == -1 || pri>=prices[stack[top]])    stack[++top] = i;
        else {  // decrese
            for (j=top; j > -1 && pri < prices[stack[j]]; j--) {
                answer[stack[j]] = i - stack[j];
                top--;
            }
            stack[++top] = i;
        }
    }
    for (i = 0 ; i <= top; i++) {
        t = stack[i];
        answer[t] = len - t-1;
    }
    return answer;
}